Core: Expired Snapshot files in a transaction should be deleted.

bartash opened the pull request:
When a snapshot is expired as part of a transaction, the snapshot file(s) should be deleted when the transaction commits. A recent change (#6634) ensured that files are not deleted when they have also been committed as part of a transaction, but this breaks the simple case where no new files are committed. Fix this by not skipping deletion when the list of committed files is empty.

Closes #9182 

TESTING:

Extended a unit test to ensure that snapshot files are deleted. Ran the test without the fix on a branch where #6634 was reverted to show that this is a regression.

bartash commented:
@amogh-jahagirdar can you please take a look when you have time as you implemented #6634

amogh-jahagirdar commented:
Thanks for the PR @bartash I'm taking a look.

amogh-jahagirdar reviewed (commented):
Thanks for the fix, this does seem like a legitimate bug. Could we add a test to `TestRemoveSnapshots` ?  I think there's no harm with the test in `TestSequenceNumberForV2Table` but `TestRemoveSnapshot` would be a better place for tests around this. There's a `deletedFiles` that's visible for testing on the transaction and I think could be used for assertions on files we expect to be deleted.

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java:
```diff
@@ -319,12 +321,18 @@ public void testExpirationInTransaction() {
     V2Assert.assertEquals(
         "Last sequence number should be 2", 2, readMetadata().lastSequenceNumber());
+    V2Assert.assertEquals(
+        "Should be 1 snapshot file as 1 was deleted",
```
Nit: Typically we use the term "manifest list" instead of  "snapshot file" could we call it that?

bartash commented:
@amogh-jahagirdar thanks for the helpful suggestions.
I pushed a new version.
Changes from version 1:
- Used the term “manifest list file” instead of “snapshot file”.
- Fixed the order of expected and actual parameters in my added assertions in TestSequenceNumberForV2Table.
- Added a new test in TestRemoveSnapshots(). The existing mechanism for tracking deleted files in TestRemoveSnapshots does not work for deletes inside transactions as the function passed in deleteWith() is called before the files are actually deleted in BaseTransaction, which is where my fix is.

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java:
```diff
@@ -309,6 +309,8 @@ public void testExpirationInTransaction() {
     V2Assert.assertEquals(
         "Last sequence number should be 1", 1, readMetadata().lastSequenceNumber());
+    V2Assert.assertEquals(
+        "Should be 1 manifest list file", 1, listManifestListFiles(table.location()).size());
```
Nit: could we just call it "manifest list" (the file is a bit redundant)

so the assertion message would be "Should be 1 manifest list"

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java:
```diff
@@ -319,12 +321,18 @@ public void testExpirationInTransaction() {
     V2Assert.assertEquals(
         "Last sequence number should be 2", 2, readMetadata().lastSequenceNumber());
+    V2Assert.assertEquals(
+        "Should be 2 manifest list files", 2, listManifestListFiles(table.location()).size());
```
Nit: "Should be 2 manifest lists"

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java:
```diff
@@ -319,12 +321,18 @@ public void testExpirationInTransaction() {
     V2Assert.assertEquals(
         "Last sequence number should be 2", 2, readMetadata().lastSequenceNumber());
+    V2Assert.assertEquals(
+        "Should be 1 manifest list file as 1 was deleted",
```
Same as above: "Should be 1 manifest list"

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/TableTestBase.java:
```diff
@@ -212,6 +212,15 @@ List<File> listManifestFiles(File tableDirToList) {
                         && Files.getFileExtension(name).equalsIgnoreCase("avro")));
   }
 
+  List<File> listManifestListFiles(String tableDirToList) {
```
Nit: listManifestLists for the function name

amogh-jahagirdar reviewed (approved):
Overall this looks good to me, just some naming nits. Thanks for the fix. cc @rdblue In case he has any comments.

bartash commented:
Thanks @amogh-jahagirdar I pushed changes for the nits.

Fokko reviewed (approved):
Once this gets in, can you create a PR against the `1.4.x` branch?

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/BaseTransaction.java:
```diff
@@ -446,20 +446,16 @@ private void commitSimpleTransaction() {
+          .onFailure((file, exc) -> LOG.warn("Failed to delete uncommitted file: {}", file, exc))
+          .run(
+              path -> {
+                if (committedFiles == null || !committedFiles.contains(path)) {
```
I just realized alternatively we could've just returned an empty set in `committedFiles` instead of null and then could've removed the  (committedFiles == null) check (in addition to the current top level check). I'm not super opinionated on that though (it can be done in a follow on)

rdblue replied:
I think this should actually change `committedFiles` to return `ImmutableSet.of()` if there are no new snapshot IDs. The logic is correct to warn if the other reason null is returned happens (a committed snapshot is missing). `null` signals that the output of the method is invalid, which we assumed was the case if there are no committed snapshots. But here we have a case where it's a valid case to have no committed snapshots and therefore no committed files.

amogh-jahagirdar replied:
I have a PR https://github.com/apache/iceberg/pull/9221 for addressing this.

amogh-jahagirdar commented:
Thanks @bartash for the fix! Merging, @rdblue @jbonofre let me know if you have any concerns

bartash commented:
> Once this gets in, can you create a PR against the `1.4.x` branch?

Clean cherry-pick to 1.4.x is in #9223
